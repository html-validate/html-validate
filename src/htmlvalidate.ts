import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
