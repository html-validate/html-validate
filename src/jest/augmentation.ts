/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
