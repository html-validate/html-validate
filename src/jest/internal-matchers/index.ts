export { toBeToken } from "./to-be-token";
