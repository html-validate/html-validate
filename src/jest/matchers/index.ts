export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
