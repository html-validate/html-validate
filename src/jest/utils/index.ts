export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
