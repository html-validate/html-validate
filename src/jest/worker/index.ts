export { type ValidateStringFn } from "./worker";
export { createSyncFn } from "./create-sync-fn";
export { default as workerPath } from "./worker.ts?worker&url";
