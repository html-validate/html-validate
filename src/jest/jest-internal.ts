import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
