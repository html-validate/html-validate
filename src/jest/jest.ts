import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
