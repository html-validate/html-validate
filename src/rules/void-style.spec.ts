import { HtmlValidate } from "../htmlvalidate";
import "../jest";

describe("rule void-style", () => {
	let htmlvalidate: HtmlValidate;

	describe("default", () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				rules: { "void-style": "error" },
			});
		});

		it("should not report when void element omitted end tag", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<input>");
			expect(report).toBeValid();
		});

		it("should report when void element is self-closed", async () => {
			expect.assertions(2);
			const report = await htmlvalidate.validateString("<input/>");
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"void-style",
				"Expected omitted end tag <input> instead of self-closing element <input/>",
			);
		});

		it("should not report when non-void element has end tag", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<div></div>");
			expect(report).toBeValid();
		});

		it("should not report when non-void element is self-closed", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<div/>");
			expect(report).toBeValid();
		});

		it("should not report when xml namespaces is used", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<xi:include/>");
			expect(report).toBeValid();
		});
	});

	describe('configured with style="omit"', () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				rules: { "void-style": ["error", { style: "omit" }] },
			});
		});

		it("should not report when void element omits end tag", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<input>");
			expect(report).toBeValid();
		});

		it("should report when void element is self-closed", async () => {
			expect.assertions(2);
			const report = await htmlvalidate.validateString("<input/>");
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"void-style",
				"Expected omitted end tag <input> instead of self-closing element <input/>",
			);
		});
	});

	describe('configured with style="selfclose"', () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				rules: { "void-style": ["error", { style: "selfclose" }] },
			});
		});

		it("should report when void element omits end tag", async () => {
			expect.assertions(2);
			const report = await htmlvalidate.validateString("<input>");
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"void-style",
				"Expected self-closing element <input/> instead of omitted end-tag <input>",
			);
		});

		it("should not report when void element is self-closed", async () => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString("<input/>");
			expect(report).toBeValid();
		});
	});

	it("should throw error if configured with invalid value", () => {
		expect.assertions(1);
		expect(() => {
			return new HtmlValidate({
				rules: { "void-style": ["error", { style: "foobar" }] },
			});
		}).toThrowErrorMatchingInlineSnapshot(
			`"Rule configuration error: /rules/void-style/1/style must be equal to one of the allowed values: omit, selfclose, selfclosing"`,
		);
	});

	it("should have documentation", async () => {
		expect.assertions(2);
		htmlvalidate = new HtmlValidate({
			rules: { "void-style": "error" },
		});
		const context1 = {
			style: 1,
			tagName: "foo",
		};
		const context2 = {
			style: 2,
			tagName: "bar",
		};
		const docs1 = await htmlvalidate.getContextualDocumentation({
			ruleId: "void-style",
			context: context1,
		});
		const docs2 = await htmlvalidate.getContextualDocumentation({
			ruleId: "void-style",
			context: context2,
		});
		expect(docs1).toMatchSnapshot();
		expect(docs2).toMatchSnapshot();
	});
});
