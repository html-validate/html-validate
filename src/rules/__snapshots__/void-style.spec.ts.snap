// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule void-style should have documentation 1`] = `
{
  "description": "The current configuration requires void elements to omit end tag, use <foo> instead.",
  "url": "https://html-validate.org/rules/void-style.html",
}
`;

exports[`rule void-style should have documentation 2`] = `
{
  "description": "The current configuration requires void elements to be self-closed, use <bar/> instead.",
  "url": "https://html-validate.org/rules/void-style.html",
}
`;
